### PR TITLE
[SPMD] Support partial replication.

### PR DESCRIPTION
**Pull request by yeounoh**
Enable partial replication where the input tensor is sharded across groups of devices (tiled) and replicated within each group (a.k.a. device subgroup). This is needed before we allow mapping lower rank tensor to a 3D mesh. cc @khatwanimohit @alanwaketan @jonb377

### CODE REVIEW

**Comment by yeounoh** [comment]
@khatwanimohit let's sync offline sometime next week, so we can discuss the 3D sharding use case.

File: test/spmd/test_xla_sharding.py [hunk 1]
```diff
@@ -50,9 +50,11 @@ def test_xla_shards(self):
...
       expected = torch.arange(start, end, dtype=torch.float32)
       self.assertTrue(torch.allclose(shard.data, expected))
-      self.assertIsInstance(shard.indices, list)
-      self.assertEqual(len(shard.indices), len(t.shape))
-      self.assertEqual(shard.indices[0], slice(start, end, 1))
+      if isinstance(shard.indices, list):
+        self.assertEqual(len(shard.indices), len(t.shape))
+        self.assertEqual(shard.indices[0], slice(start, end, 1))
+      else:
+        self.assertIsInstance(shard.indices, type(Ellipsis))
```
Comment (jonb377) [inline test/spmd/test_xla_sharding.py:57]: Ah sharding across one device is `MAXIMAL` which becomes `REPLICATED`, so this happens on CPU 👍

File: test/spmd/test_xla_sharding.py [hunk 2]
```diff
@@ -131,15 +137,61 @@ def test_mark_sharding_4d(self):
...
+    if self.n_devices > 1:
+      annotation = '{devices=[1,1,%d,%d]%s}' % (
+          z_dim, self.n_devices // z_dim, ','.join(
+              [str(i) for i in range(self.n_devices)]))
+      self.assertEqual(annotation, torch_xla._XLAC._get_xla_sharding_spec(xt))
 
     actual = (xt + xt).cpu()
     self.assertTrue(torch.allclose(expected, actual))
 
+  def test_mark_sharding_partial(self):
```
Comment (jonb377) [inline test/spmd/test_xla_sharding.py:150]: Can we also test local_shards under partial replication?

File: torch_xla/csrc/xla_sharding_util.cpp [hunk 3]
```diff
@@ -332,6 +405,13 @@ ShardingUtil::GetShardIndicesForDevices(
       int offset = i;
       std::vector<at::indexing::TensorIndex> indices;
       for (int j = tile_shape.size() - 1; j >= 0; j--) {
+        if (sharding.replicate_on_last_tile_dim() &&
+            j == tile_shape.size() - 1) {
+          // the last tile assignment dimension is replicated, which implies
+          // that the consecutive `tile_shape[j]` devices hold the replicated.
+          offset /= tile_shape[j];
+          continue;
+        }
```
Comment (jonb377) [inline torch_xla/csrc/xla_sharding_util.cpp:414]: High-level question - how does `tile_shape` change when the tensor is partially replicated?

Right now, this logic makes sense to me if partial replication adds a dimension to track replica group and `tile_shape[j] == 1` on the replicated dimensions, is that a correct understanding?

Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:414]: That's correct.

File: test/spmd/test_xla_sharding.py (hunk 2 above)
Comment (yeounoh) [reply test/spmd/test_xla_sharding.py:150]: +1 yes, we could with the new shards access API now.

File: torch_xla/csrc/xla_sharding_util.cpp [hunk 4]
```diff
@@ -36,7 +43,7 @@ std::vector<int64_t> TileAssignmentDimensions(
     XLA_CHECK(r.size() > 0)
         << "Invalid argument: empty list is not a valid element type.";
     dims.push_back(r.size());
-    auto type = r[0].attr("__class__").attr("__name__").cast<std::string>();
+    std::string type = GetPyType(r[0]);
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:46]: I like it, let me take a look.

**Review by jonb377 — APPROVED** [review]
LGTM, thanks Yeounoh! 🚀

**Comment by yeounoh** [comment]
Thanks for the comments, I will have a follow-up PR, since they are minor (+ i dont' have access to tpuvm at the moment... to test)

File: torch_xla/experimental/xla_sharding.py [hunk 5]
```diff
@@ -71,6 +72,59 @@ def get_logical_mesh(self):
...
+  elif any(d is None for d in partition_spec):
+    sharding_type = ShardingType.PARTIAL
+  return sharding_type
+
+
+def _get_tile_assignment(mesh: Mesh) -> List[int]:
+  return mesh.get_logical_mesh().tolist()
+
+
+def _get_group_assignment(
```
Comment (alanwaketan) [inline torch_xla/experimental/xla_sharding.py:101]: Can you explain a bit on what this helper does?

**Pull request merged**